app/models/category.rb

```diff
@@ -227,6 +225,32 @@ class Category < ActiveRecord::Base
   # Allows us to skip creating the category definition topic in tests.
   attr_accessor :skip_category_definition
 
+  def self.preload_user_fields!(guardian, categories)
+    # Load notification levels
+    notification_levels = CategoryUser.notification_levels_for(guardian.user)
+    notification_levels.default = CategoryUser.default_notification_level
+
+    # Load permissions
+    allowed_topic_create_ids =
+      if !guardian.is_admin? && !guardian.is_anonymous?
+        Category.topic_create_allowed(guardian).where(id: categories.map(&:id)).pluck(:id).to_set
```

Comment:
Am curious why `to_set` is necessary here when I don't think the query will generate duplicated rows. If it is, we could just ensure we pluck "DISTINCT id" instead.

Reply:
This is for performance reasons. Set is O(1) complexity, while looking up in a list is O(n).

Reply:
That is nice. I think we should perhaps leave a comment here and below in case someone comes across this in the future.